lib/robots/dor_repo/assembly/jp2_create.rb

```diff
@@ -24,82 +24,124 @@ def perform_work
 
         # For each supported image type that is part of specific resource types, generate a jp2 derivative
         # and modify structural metadata to reflect the new file.
-        # grab all the file node tuples for each valid resource type that we want to generate derivates for
         # rubocop:disable Metrics/PerceivedComplexity
         def create_jp2s(assembly_item, cocina_model)
           logger.info("Creating JP2s for #{assembly_item.druid.id}")
           file_sets = cocina_model.structural.to_h.fetch(:contains) # make this a mutable hash
           file_sets.each do |file_set|
-            next unless [Cocina::Models::FileSetType.page, Cocina::Models::FileSetType.image].include?(file_set.fetch(:type))
+            next if skip_fileset?(file_set)
 
-            files = file_set.dig(:structural, :contains)
+            cocina_files = file_set.dig(:structural, :contains)
 
-            tuples_to_add = []
-            files.each do |file|
-              filepath = assembly_item.path_finder.path_to_content_file(file.fetch(:filename))
+            new_cocina_files = cocina_files.dup
+            cocina_files.each do |cocina_file|
+              # If the file cocina file has a mimetype and it is not jp2able, skip it
+              next if skip_cocina_file?(cocina_file)
 
-              # Allow for files that are not changing, hence are not present.
-              next unless File.exist?(filepath)
+              filename = cocina_file.fetch(:filename)
+              filepath = filepath_for(filename, assembly_item)
 
-              object_file = ::Assembly::ObjectFile.new(filepath)
+              # If the file exists and it is not jp2able, skip it
+              next if skip_file?(filepath)
 
-              next unless object_file.jp2able?
+              assembly_image = assembly_image_for(filepath)
+              jp2_filename = jp2_filename_for(filename, assembly_image)
+              jp2_filepath = filepath_for(jp2_filename, assembly_item)
 
-              img = ::Assembly::Image.new(object_file.path)
-              tuples_to_add << [file, img]
-            end
+              cocina_jp2_file = find_cocina_jp2_file(cocina_files)
 
-            if tuples_to_add.present?
-              # Remove existing jp2 nodes
-              files.delete_if { |file| file.fetch(:filename).ends_with?('.jp2') }
-            end
+              # If the file exists and there is a jp2 cocina file, skip it
+              next if File.exist?(filepath) && File.exist?(jp2_filepath) && cocina_jp2_file.present?
+
+              # If the file does not exist and there is a jp2 cocina file, skip it
+              next if !File.exist?(filepath) && cocina_jp2_file.present?
```

Comment:
would it make sense to check for the cocina_jp2 file before some of these steps?

Reply:
It's another option, that I'm not sure that any formulation of this logic is going to be clearer than any other. If you had a formulation, I'd be happy to consider.

Reply:
backing away slowly.  There might be a way but I'm not going to spend the time on it.  If someone wants, they can refactor it later.  Working code wins.